<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>VDJMLpy &mdash; vdjml documentation</title>
    
    <link rel="stylesheet" href="_static/sphinx_vdjml.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="vdjml documentation" href="index.html" />
    <link rel="next" title="Building vdjml" href="build.html" />
    <link rel="prev" title="vdjml" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="build.html" title="Building vdjml"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="vdjml"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">vdjml</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="vdjmlpy">
<h1>VDJMLpy<a class="headerlink" href="#vdjmlpy" title="Permalink to this headline">¶</a></h1>
<p>VDJMLpy is a Python module for working with the results of immune receptor
sequence alignment in <a class="reference external" href="xsd_doc/1/index.html">VDJML</a> format.
It is built as bindings to libVDJML, a C++ library.</p>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><em>Index</em></a></li>
<li><a class="reference internal" href="search.html"><em>Search Page</em></a></li>
</ul>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p class="graphviz">
<img src="_images/graphviz-270aa5e4589186972f99681dd7aa365e86d684bb.png" alt="vdjml_py Overview" usemap="#VDJML_py" />
<map id="VDJML_py" name="VDJML_py">
<area shape="rect" id="node1" href="#vdjml._vdjml_py.Results_meta" target="_top" title="Results_meta" alt="" coords="157,272,283,357"/>
<area shape="rect" id="node8" href="#vdjml._vdjml_py.Result_store" target="_top" title="Result_store" alt="" coords="253,405,371,491"/>
<area shape="rect" id="node2" href="#vdjml._vdjml_py.Read_result" target="_top" title="Read_result" alt="" coords="349,272,460,357"/>
<area shape="rect" id="node3" href="#vdjml._vdjml_py.Segment_match" target="_top" title="Segment_match" alt="" coords="213,139,360,224"/>
<area shape="rect" id="node4" href="#vdjml._vdjml_py.Aa_substitution" target="_top" title="Aa_substitution" alt="" coords="5,5,147,91"/>
<area shape="rect" id="node5" href="#vdjml._vdjml_py.Gl_segment_match" target="_top" title="Gl_segment_match" alt="" coords="215,5,385,91"/>
<area shape="rect" id="node6" href="#vdjml._vdjml_py.Segment_combination" target="_top" title="Segment_combination" alt="" coords="428,139,623,224"/>
<area shape="rect" id="node7" href="#vdjml._vdjml_py.Gene_region" target="_top" title="Gene_region" alt="" coords="466,5,584,91"/>
</map>
</p>
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
</div>
<div class="section" id="module-vdjml">
<span id="api-reference"></span><h2>API reference<a class="headerlink" href="#module-vdjml" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-vdjml._vdjml_py"></span><dl class="class">
<dt id="vdjml._vdjml_py.Aa_substitution">
<em class="property">class </em><tt class="descname">Aa_substitution</tt><big>(</big><em>(object)arg1</em>, <em>(int)read_pos0</em>, <em>(Aminoacid)read_aa</em>, <em>(Aminoacid)gl_aa</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Aa_substitution" title="Permalink to this definition">¶</a></dt>
<dd><p>amino acid substitution</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Aa_substitution.gl_aa">
<tt class="descname">gl_aa</tt><big>(</big><em>(Aa_substitution)arg1</em><big>)</big> &rarr; Aminoacid :<a class="headerlink" href="#vdjml._vdjml_py.Aa_substitution.gl_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>amino acid encoded by the germline sequence</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aa_substitution.read_aa">
<tt class="descname">read_aa</tt><big>(</big><em>(Aa_substitution)arg1</em><big>)</big> &rarr; Aminoacid :<a class="headerlink" href="#vdjml._vdjml_py.Aa_substitution.read_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>amino acid encoded by the read sequence</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aa_substitution.read_position">
<tt class="descname">read_position</tt><big>(</big><em>(Aa_substitution)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Aa_substitution.read_position" title="Permalink to this definition">¶</a></dt>
<dd><p>0-based read position of the codone&#8217;s first nucleotide</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Aa_substitutions_set">
<em class="property">class </em><tt class="descname">Aa_substitutions_set</tt><a class="headerlink" href="#vdjml._vdjml_py.Aa_substitutions_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Set of amino acid substitutions</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Aa_substitutions_set.empty">
<tt class="descname">empty</tt><big>(</big><em>(Aa_substitutions_set)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Aa_substitutions_set.empty" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Aligner_id">
<em class="property">class </em><tt class="descname">Aligner_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Aligner_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligner software ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Aligner_info">
<em class="property">class </em><tt class="descname">Aligner_info</tt><big>(</big><em>(object)arg1</em>, <em>(str)name</em>, <em>(str)version</em><span class="optional">[</span>, <em>(str)parameters=''</em><span class="optional">[</span>, <em>(str)uri=''</em><span class="optional">[</span>, <em>(int)run_id=0</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Aligner_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Info about aligner software</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_info.id">
<tt class="descname">id</tt><big>(</big><em>(Aligner_info)arg1</em><big>)</big> &rarr; Aligner_id :<a class="headerlink" href="#vdjml._vdjml_py.Aligner_info.id" title="Permalink to this definition">¶</a></dt>
<dd><p>aligner software ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_info.name">
<tt class="descname">name</tt><big>(</big><em>(Aligner_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Aligner_info.name" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_info.parameters">
<tt class="descname">parameters</tt><big>(</big><em>(Aligner_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Aligner_info.parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_info.run_id">
<tt class="descname">run_id</tt><big>(</big><em>(Aligner_info)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Aligner_info.run_id" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_info.uri">
<tt class="descname">uri</tt><big>(</big><em>(Aligner_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Aligner_info.uri" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_info.version">
<tt class="descname">version</tt><big>(</big><em>(Aligner_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Aligner_info.version" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Aligner_map">
<em class="property">class </em><tt class="descname">Aligner_map</tt><a class="headerlink" href="#vdjml._vdjml_py.Aligner_map" title="Permalink to this definition">¶</a></dt>
<dd><p>aligner info map</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Aligner_map.empty">
<tt class="descname">empty</tt><big>(</big><em>(Aligner_map)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Aligner_map.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Btop">
<em class="property">class </em><tt class="descname">Btop</tt><big>(</big><em>(object)arg1</em>, <em>(str)arg2</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Btop" title="Permalink to this definition">¶</a></dt>
<dd><p>Blast trace-back operations, alignment description</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Btop.empty">
<tt class="descname">empty</tt><big>(</big><em>(Btop)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Btop.empty" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if Btop is empty</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Btop_stats">
<em class="property">class </em><tt class="descname">Btop_stats</tt><big>(</big><em>(object)arg1</em>, <em>(Btop)arg2</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Btop_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Btop_statistics</p>
<dl class="attribute">
<dt id="vdjml._vdjml_py.Btop_stats.deletions_">
<tt class="descname">deletions_</tt><a class="headerlink" href="#vdjml._vdjml_py.Btop_stats.deletions_" title="Permalink to this definition">¶</a></dt>
<dd><p>number of deletions (from read sequence)</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Btop_stats.gl_len_">
<tt class="descname">gl_len_</tt><a class="headerlink" href="#vdjml._vdjml_py.Btop_stats.gl_len_" title="Permalink to this definition">¶</a></dt>
<dd><p>length of the aligned germline sequence</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Btop_stats.insertions_">
<tt class="descname">insertions_</tt><a class="headerlink" href="#vdjml._vdjml_py.Btop_stats.insertions_" title="Permalink to this definition">¶</a></dt>
<dd><p>number of insertions (in read sequence)</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Btop_stats.matches">
<tt class="descname">matches</tt><big>(</big><em>(Btop_stats)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Btop_stats.matches" title="Permalink to this definition">¶</a></dt>
<dd><p>number of matches in the alignment</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Btop_stats.read_len_">
<tt class="descname">read_len_</tt><a class="headerlink" href="#vdjml._vdjml_py.Btop_stats.read_len_" title="Permalink to this definition">¶</a></dt>
<dd><p>length of the aligned read sequence</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Btop_stats.substitutions_">
<tt class="descname">substitutions_</tt><a class="headerlink" href="#vdjml._vdjml_py.Btop_stats.substitutions_" title="Permalink to this definition">¶</a></dt>
<dd><p>number of substitutions</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Codon_match">
<em class="property">class </em><tt class="descname">Codon_match</tt><a class="headerlink" href="#vdjml._vdjml_py.Codon_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Information about a pair of aligned codons, nucleotide triples, which may potentially contain gaps</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.gl_char">
<tt class="descname">gl_char</tt><big>(</big><em>(Codon_match)arg1</em>, <em>(int)i</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.gl_char" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i</strong> (<em>int</em>) &#8211; nucleotide position in the codon [0,2]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">codon nucleotide in germline sequence</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.gl_nuc">
<tt class="descname">gl_nuc</tt><big>(</big><em>(Codon_match)arg1</em>, <em>(int)i</em><big>)</big> &rarr; Nucleotide :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.gl_nuc" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i</strong> (<em>int</em>) &#8211; nucleotide position in the codon [0,2]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">codon nucleotide in germline sequence</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.gl_pos">
<tt class="descname">gl_pos</tt><big>(</big><em>(Codon_match)arg1</em><span class="optional">[</span>, <em>(int)i=0</em><span class="optional">]</span><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.gl_pos" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i</strong> (<em>int</em>) &#8211; nucleotide position in the codon [0,2]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">position of codon nucleotide in germline sequence; if a deletion, return position of the first nucleotide to the right, if at end of sequence, return sequence length</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_gl_contiguous">
<tt class="descname">is_gl_contiguous</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_gl_contiguous" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if germline codon contains no gaps</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_gl_translatable">
<tt class="descname">is_gl_translatable</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_gl_translatable" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if germline codon can be unambiguously translated</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_match">
<tt class="descname">is_match</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_match" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if same nucleotides in read and germline</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_read_contiguous">
<tt class="descname">is_read_contiguous</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_read_contiguous" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if read codon contains no gaps</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_read_translatable">
<tt class="descname">is_read_translatable</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_read_translatable" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if read codon can be unambiguously translated</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_silent">
<tt class="descname">is_silent</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_silent" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.is_match" title="vdjml._vdjml_py.Codon_match.is_match"><tt class="xref py py-meth docutils literal"><span class="pre">is_match()</span></tt></a> or if <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.translate_read" title="vdjml._vdjml_py.Codon_match.translate_read"><tt class="xref py py-meth docutils literal"><span class="pre">translate_read()</span></tt></a> == <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.translate_gl" title="vdjml._vdjml_py.Codon_match.translate_gl"><tt class="xref py py-meth docutils literal"><span class="pre">translate_gl()</span></tt></a></td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Raises RuntimeError:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">if not <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.is_match" title="vdjml._vdjml_py.Codon_match.is_match"><tt class="xref py py-meth docutils literal"><span class="pre">is_match()</span></tt></a> and not <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.is_translatable" title="vdjml._vdjml_py.Codon_match.is_translatable"><tt class="xref py py-meth docutils literal"><span class="pre">is_translatable()</span></tt></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.is_translatable">
<tt class="descname">is_translatable</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.is_translatable" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if both codons can be unambiguously translated</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.read_char">
<tt class="descname">read_char</tt><big>(</big><em>(Codon_match)arg1</em>, <em>(int)i</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.read_char" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i</strong> (<em>int</em>) &#8211; nucleotide position in the codon [0,2]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">codon nucleotide in read sequence</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.read_nuc">
<tt class="descname">read_nuc</tt><big>(</big><em>(Codon_match)arg1</em>, <em>(int)i</em><big>)</big> &rarr; Nucleotide :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.read_nuc" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i</strong> (<em>int</em>) &#8211; nucleotide position in the codon [0,2]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">codon nucleotide in read sequence</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.read_pos">
<tt class="descname">read_pos</tt><big>(</big><em>(Codon_match)arg1</em><span class="optional">[</span>, <em>(int)i=0</em><span class="optional">]</span><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.read_pos" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i</strong> (<em>int</em>) &#8211; nucleotide position in the codon [0,2]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">position of codon nucleotide in read sequence; if a deletion, return position of the first nucleotide to the right, if at end of sequence, return sequence length</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.translate_gl">
<tt class="descname">translate_gl</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; Aminoacid :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.translate_gl" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">amino acid encoded by germline codon</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Raises RuntimeError:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">if not <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.is_gl_translatable" title="vdjml._vdjml_py.Codon_match.is_gl_translatable"><tt class="xref py py-meth docutils literal"><span class="pre">is_gl_translatable()</span></tt></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Codon_match.translate_read">
<tt class="descname">translate_read</tt><big>(</big><em>(Codon_match)arg1</em><big>)</big> &rarr; Aminoacid :<a class="headerlink" href="#vdjml._vdjml_py.Codon_match.translate_read" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">amino acid encoded by read codon</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Raises RuntimeError:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">if not <a class="reference internal" href="#vdjml._vdjml_py.Codon_match.is_read_translatable" title="vdjml._vdjml_py.Codon_match.is_read_translatable"><tt class="xref py py-meth docutils literal"><span class="pre">is_read_translatable()</span></tt></a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gene_region">
<em class="property">class </em><tt class="descname">Gene_region</tt><big>(</big><em>(object)arg1</em>, <em>(Region_id)region</em>, <em>(Numsys_id)num_system</em>, <em>(Aligner_id)aligner</em>, <em>(Interval)range</em>, <em>(Match_metrics)mm</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gene_region" title="Permalink to this definition">¶</a></dt>
<dd><p>information about gene region alignment</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region.aligner">
<tt class="descname">aligner</tt><big>(</big><em>(Gene_region)arg1</em><big>)</big> &rarr; Aligner_id :<a class="headerlink" href="#vdjml._vdjml_py.Gene_region.aligner" title="Permalink to this definition">¶</a></dt>
<dd><p>aligner ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region.match_metrics">
<tt class="descname">match_metrics</tt><big>(</big><em>(Gene_region)arg1</em><big>)</big> &rarr; Match_metrics :<a class="headerlink" href="#vdjml._vdjml_py.Gene_region.match_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>match metrics</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region.numbering_system">
<tt class="descname">numbering_system</tt><big>(</big><em>(Gene_region)arg1</em><big>)</big> &rarr; Numsys_id :<a class="headerlink" href="#vdjml._vdjml_py.Gene_region.numbering_system" title="Permalink to this definition">¶</a></dt>
<dd><p>numbering system ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region.read_range">
<tt class="descname">read_range</tt><big>(</big><em>(Gene_region)arg1</em><big>)</big> &rarr; Interval :<a class="headerlink" href="#vdjml._vdjml_py.Gene_region.read_range" title="Permalink to this definition">¶</a></dt>
<dd><p>read sequence range</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region.region_type">
<tt class="descname">region_type</tt><big>(</big><em>(Gene_region)arg1</em><big>)</big> &rarr; Region_id :<a class="headerlink" href="#vdjml._vdjml_py.Gene_region.region_type" title="Permalink to this definition">¶</a></dt>
<dd><p>region ID</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gene_region_map">
<em class="property">class </em><tt class="descname">Gene_region_map</tt><a class="headerlink" href="#vdjml._vdjml_py.Gene_region_map" title="Permalink to this definition">¶</a></dt>
<dd><p>map of gene region names</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region_map.empty">
<tt class="descname">empty</tt><big>(</big><em>(Gene_region_map)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Gene_region_map.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gene_region_set">
<em class="property">class </em><tt class="descname">Gene_region_set</tt><a class="headerlink" href="#vdjml._vdjml_py.Gene_region_set" title="Permalink to this definition">¶</a></dt>
<dd><p>set of gene regions</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gene_region_set.empty">
<tt class="descname">empty</tt><big>(</big><em>(Gene_region_set)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Gene_region_set.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Germline_db_map">
<em class="property">class </em><tt class="descname">Germline_db_map</tt><a class="headerlink" href="#vdjml._vdjml_py.Germline_db_map" title="Permalink to this definition">¶</a></dt>
<dd><p>map of germline database descriptions</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Germline_db_map.empty">
<tt class="descname">empty</tt><big>(</big><em>(Germline_db_map)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Germline_db_map.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_db_id">
<em class="property">class </em><tt class="descname">Gl_db_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gl_db_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Germline database ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_db_info">
<em class="property">class </em><tt class="descname">Gl_db_info</tt><big>(</big><em>(object)arg1</em>, <em>(str)name</em>, <em>(str)version</em>, <em>(str)species</em>, <em>(str)url</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gl_db_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Info about germline sequences database</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gl_db_info.id">
<tt class="descname">id</tt><big>(</big><em>(Gl_db_info)arg1</em><big>)</big> &rarr; Gl_db_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_db_info.id" title="Permalink to this definition">¶</a></dt>
<dd><p>germline database ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_db_info.name">
<tt class="descname">name</tt><big>(</big><em>(Gl_db_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Gl_db_info.name" title="Permalink to this definition">¶</a></dt>
<dd><p>germline database name</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_db_info.species">
<tt class="descname">species</tt><big>(</big><em>(Gl_db_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Gl_db_info.species" title="Permalink to this definition">¶</a></dt>
<dd><p>germline database species</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_db_info.uri">
<tt class="descname">uri</tt><big>(</big><em>(Gl_db_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Gl_db_info.uri" title="Permalink to this definition">¶</a></dt>
<dd><p>germline database URI</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_db_info.version">
<tt class="descname">version</tt><big>(</big><em>(Gl_db_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Gl_db_info.version" title="Permalink to this definition">¶</a></dt>
<dd><p>germline database version</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_seg_id">
<em class="property">class </em><tt class="descname">Gl_seg_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gl_seg_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Germline segment ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_seg_match_id">
<em class="property">class </em><tt class="descname">Gl_seg_match_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gl_seg_match_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Germline segment match ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_segment_info">
<em class="property">class </em><tt class="descname">Gl_segment_info</tt><big>(</big><em>(object)arg1</em>, <em>(Gl_db_id)gl_db_id</em>, <em>(str)vdj</em>, <em>(str)name</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_info" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment description</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_info.gl_database">
<tt class="descname">gl_database</tt><big>(</big><em>(Gl_segment_info)arg1</em><big>)</big> &rarr; Gl_db_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_info.gl_database" title="Permalink to this definition">¶</a></dt>
<dd><p>germline database ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_info.id">
<tt class="descname">id</tt><big>(</big><em>(Gl_segment_info)arg1</em><big>)</big> &rarr; Gl_seg_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_info.id" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_info.name">
<tt class="descname">name</tt><big>(</big><em>(Gl_segment_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_info.name" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment name</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_info.segment_type">
<tt class="descname">segment_type</tt><big>(</big><em>(Gl_segment_info)arg1</em><big>)</big> &rarr; object :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_info.segment_type" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment type</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_segment_map">
<em class="property">class </em><tt class="descname">Gl_segment_map</tt><a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Map of germline segments aligned to read interval</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_map.empty">
<tt class="descname">empty</tt><big>(</big><em>(Gl_segment_map)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_map.empty" title="Permalink to this definition">¶</a></dt>
<dd><p>todo:</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Gl_segment_match">
<em class="property">class </em><tt class="descname">Gl_segment_match</tt><big>(</big><em>(object)arg1</em>, <em>(Numsys_id)num_system</em>, <em>(Aligner_id)aligner</em>, <em>(Gl_seg_id)germline_segment</em>, <em>(int)gl_pos0</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment to germline segment</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_match.aligner">
<tt class="descname">aligner</tt><big>(</big><em>(Gl_segment_match)arg1</em><big>)</big> &rarr; Aligner_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_match.aligner" title="Permalink to this definition">¶</a></dt>
<dd><p>ID of the software that aligned the germline segment</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_match.gl_position">
<tt class="descname">gl_position</tt><big>(</big><em>(Gl_segment_match)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_match.gl_position" title="Permalink to this definition">¶</a></dt>
<dd><p>first aligned position index of the germline segment</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_match.gl_segment">
<tt class="descname">gl_segment</tt><big>(</big><em>(Gl_segment_match)arg1</em><big>)</big> &rarr; Gl_seg_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_match.gl_segment" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_match.id">
<tt class="descname">id</tt><big>(</big><em>(Gl_segment_match)arg1</em><big>)</big> &rarr; Gl_seg_match_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_match.id" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment match ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Gl_segment_match.num_system">
<tt class="descname">num_system</tt><big>(</big><em>(Gl_segment_match)arg1</em><big>)</big> &rarr; Numsys_id :<a class="headerlink" href="#vdjml._vdjml_py.Gl_segment_match.num_system" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment numbering system ID</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Interval">
<em class="property">class </em><tt class="descname">Interval</tt><a class="headerlink" href="#vdjml._vdjml_py.Interval" title="Permalink to this definition">¶</a></dt>
<dd><p>sequence interval</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Interval.first_last0">
<tt class="descname">first_last0</tt><big>(</big><em>(int)first0</em>, <em>(int)last0</em><big>)</big> &rarr; Interval :<a class="headerlink" href="#vdjml._vdjml_py.Interval.first_last0" title="Permalink to this definition">¶</a></dt>
<dd><p>create interval from 0-based first and last positions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.first_last1">
<tt class="descname">first_last1</tt><big>(</big><em>(int)first1</em>, <em>(int)last1</em><big>)</big> &rarr; Interval :<a class="headerlink" href="#vdjml._vdjml_py.Interval.first_last1" title="Permalink to this definition">¶</a></dt>
<dd><p>create interval from 1-based first and last positions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.last0">
<tt class="descname">last0</tt><big>(</big><em>(Interval)arg1</em><big>)</big> &rarr; int<a class="headerlink" href="#vdjml._vdjml_py.Interval.last0" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.last1">
<tt class="descname">last1</tt><big>(</big><em>(Interval)arg1</em><big>)</big> &rarr; int<a class="headerlink" href="#vdjml._vdjml_py.Interval.last1" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.length">
<tt class="descname">length</tt><big>(</big><em>(Interval)arg1</em><big>)</big> &rarr; int<a class="headerlink" href="#vdjml._vdjml_py.Interval.length" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.pos0">
<tt class="descname">pos0</tt><big>(</big><em>(Interval)arg1</em><big>)</big> &rarr; int<a class="headerlink" href="#vdjml._vdjml_py.Interval.pos0" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.pos0_len">
<tt class="descname">pos0_len</tt><big>(</big><em>(int)pos0</em>, <em>(int)len</em><big>)</big> &rarr; Interval :<a class="headerlink" href="#vdjml._vdjml_py.Interval.pos0_len" title="Permalink to this definition">¶</a></dt>
<dd><p>create interval from 0-based starting position and length</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Interval.pos1">
<tt class="descname">pos1</tt><big>(</big><em>(Interval)arg1</em><big>)</big> &rarr; int<a class="headerlink" href="#vdjml._vdjml_py.Interval.pos1" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Match_metrics">
<em class="property">class </em><tt class="descname">Match_metrics</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Match_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Metrics of sequence alignment</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.deletions">
<tt class="descname">deletions</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.deletions" title="Permalink to this definition">¶</a></dt>
<dd><p>number of bases deleted in the read</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.frame_shift">
<tt class="descname">frame_shift</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.frame_shift" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if a frame shift is present (out_frame_indel() || out_frame_vdj())</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.identity">
<tt class="descname">identity</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; Percent :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.identity" title="Permalink to this definition">¶</a></dt>
<dd><p>percent identity</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.insertions">
<tt class="descname">insertions</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.insertions" title="Permalink to this definition">¶</a></dt>
<dd><p>number of bases inserted in the read</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.inverted">
<tt class="descname">inverted</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.inverted" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if sequence inverted</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.mutated_invariant">
<tt class="descname">mutated_invariant</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.mutated_invariant" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if invariant amino acid mutated</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.out_frame_indel">
<tt class="descname">out_frame_indel</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.out_frame_indel" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if INDEL causes frameshift</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.out_frame_vdj">
<tt class="descname">out_frame_vdj</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.out_frame_vdj" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if VDJ rearrangement is out of frame</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.productive">
<tt class="descname">productive</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.productive" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if sequence is productive ( ! (out_frame_indel() || out_frame_vdj() || mutated_invariant()) )</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.score">
<tt class="descname">score</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.score" title="Permalink to this definition">¶</a></dt>
<dd><p>alignment score</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.stop_codon">
<tt class="descname">stop_codon</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.stop_codon" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if stop codon present</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Match_metrics.substitutions">
<tt class="descname">substitutions</tt><big>(</big><em>(Match_metrics)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Match_metrics.substitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>number of base substitutions</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Nucleotide_match">
<em class="property">class </em><tt class="descname">Nucleotide_match</tt><a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Information about a pair of aligned nucleotides, which may be a match, substitution, insertion, or deletion</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.gl_nuc">
<tt class="descname">gl_nuc</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.gl_nuc" title="Permalink to this definition">¶</a></dt>
<dd><p>return nucleotide character in germline sequence; if the mismatch is an insertion, return &#8216;-&#8216;</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.gl_pos">
<tt class="descname">gl_pos</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.gl_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>return position of nucleotide in germline sequence; if an insertion, return position of the first nucleotide to the right, if at end of sequence, return sequence length</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.is_deletion">
<tt class="descname">is_deletion</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.is_deletion" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if a deletion (from read sequence)</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.is_insertion">
<tt class="descname">is_insertion</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.is_insertion" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if insertion (in read sequence)</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.is_match">
<tt class="descname">is_match</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.is_match" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if mismatch</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.read_nuc">
<tt class="descname">read_nuc</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.read_nuc" title="Permalink to this definition">¶</a></dt>
<dd><p>return nucleotide character in read sequence; if the mismatch is a deletion, return &#8216;-&#8216;</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Nucleotide_match.read_pos">
<tt class="descname">read_pos</tt><big>(</big><em>(Nucleotide_match)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Nucleotide_match.read_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>return position of nucleotide in read sequence; if a deletion, return position of the first nucleotide to the right, if at end of sequence, return sequence length</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Num_system_map">
<em class="property">class </em><tt class="descname">Num_system_map</tt><a class="headerlink" href="#vdjml._vdjml_py.Num_system_map" title="Permalink to this definition">¶</a></dt>
<dd><p>map of numbering system names</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Num_system_map.empty">
<tt class="descname">empty</tt><big>(</big><em>(Num_system_map)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Num_system_map.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Numsys_id">
<em class="property">class </em><tt class="descname">Numsys_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Numsys_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Numbering system ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Read_result">
<em class="property">class </em><tt class="descname">Read_result</tt><big>(</big><em>(object)arg1</em>, <em>(str)arg2</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Read_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis results for one sequencing read</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Read_result.id">
<tt class="descname">id</tt><big>(</big><em>(Read_result)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Read_result.id" title="Permalink to this definition">¶</a></dt>
<dd><p>read ID string</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Read_result.insert">
<tt class="descname">insert</tt><big>(</big><em>(Read_result)arg1</em>, <em>(Segment_match)arg2</em><big>)</big> &rarr; Seg_match_id :<a class="headerlink" href="#vdjml._vdjml_py.Read_result.insert" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>insert segment match</div></blockquote>
<dl class="docutils">
<dt>insert( (Read_result)arg1, (Segment_combination)arg2) -&gt; None :</dt>
<dd>insert combination of segment matches</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Read_result.segment_combinations">
<tt class="descname">segment_combinations</tt><big>(</big><em>(Read_result)arg1</em><big>)</big> &rarr; Segment_combinations_list :<a class="headerlink" href="#vdjml._vdjml_py.Read_result.segment_combinations" title="Permalink to this definition">¶</a></dt>
<dd><p>list of segment combinations</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Read_result.segment_matches">
<tt class="descname">segment_matches</tt><big>(</big><em>(Read_result)arg1</em><big>)</big> &rarr; Segment_match_map :<a class="headerlink" href="#vdjml._vdjml_py.Read_result.segment_matches" title="Permalink to this definition">¶</a></dt>
<dd><p>map of segment matches</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Region_id">
<em class="property">class </em><tt class="descname">Region_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Region_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene region type ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Result_builder">
<em class="property">class </em><tt class="descname">Result_builder</tt><big>(</big><em>(object)arg1</em>, <em>(Results_meta)meta</em>, <em>(str)read_id</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Result_builder" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct alignment results for one sequencing read</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Result_builder.get">
<tt class="descname">get</tt><big>(</big><em>(Result_builder)arg1</em><big>)</big> &rarr; Read_result :<a class="headerlink" href="#vdjml._vdjml_py.Result_builder.get" title="Permalink to this definition">¶</a></dt>
<dd><p>get result object (internal reference)</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_builder.insert_segment_combination">
<tt class="descname">insert_segment_combination</tt><big>(</big><em>(Result_builder)arg1</em>, <em>(Seg_match_id)seg_match_1</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_2=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb9ec0&gt;</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_3=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb9f30&gt;</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_4=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb9fa0&gt;</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_5=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23eccc050&gt;</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; Segment_combination_builder<a class="headerlink" href="#vdjml._vdjml_py.Result_builder.insert_segment_combination" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_builder.insert_segment_match">
<tt class="descname">insert_segment_match</tt><big>(</big><em>(Result_builder)arg1</em>, <em>(int)read_pos0</em>, <em>(str)btop</em>, <em>(str)vdj</em>, <em>(str)seg_name</em>, <em>(int)gl_pos0</em><span class="optional">[</span>, <em>(Match_metrics)metric=&lt;vdjml._vdjml_py.Match_metrics object at 0x7fa23ecc7488&gt;</em><span class="optional">[</span>, <em>(Gl_db_id)gl_database=&lt;vdjml._vdjml_py.Gl_db_id object at 0x7fa23ecb9d70&gt;</em><span class="optional">[</span>, <em>(Numsys_id)num_system=&lt;vdjml._vdjml_py.Numsys_id object at 0x7fa23ecb9de0&gt;</em><span class="optional">[</span>, <em>(Aligner_id)aligner=&lt;vdjml._vdjml_py.Aligner_id object at 0x7fa23ecb9e50&gt;</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; Segment_match_builder :<a class="headerlink" href="#vdjml._vdjml_py.Result_builder.insert_segment_match" title="Permalink to this definition">¶</a></dt>
<dd><p>add new segment match</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_builder.release">
<tt class="descname">release</tt><big>(</big><em>(Result_builder)arg1</em><big>)</big> &rarr; Read_result :<a class="headerlink" href="#vdjml._vdjml_py.Result_builder.release" title="Permalink to this definition">¶</a></dt>
<dd><p>get final result object (independent copy); Result_builder object cannot be used anymore</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Result_factory">
<em class="property">class </em><tt class="descname">Result_factory</tt><big>(</big><em>(object)arg1</em>, <em>(Results_meta)arg2</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Result_factory" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct alignment results for many sequencing reads</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Result_factory.new_result">
<tt class="descname">new_result</tt><big>(</big><em>(Result_factory)arg1</em>, <em>(str)read_id</em><big>)</big> &rarr; Result_builder :<a class="headerlink" href="#vdjml._vdjml_py.Result_factory.new_result" title="Permalink to this definition">¶</a></dt>
<dd><p>new result builder</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_factory.set_default_aligner">
<tt class="descname">set_default_aligner</tt><big>(</big><em>(Result_factory)arg1</em>, <em>(Aligner_id)arg2</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Result_factory.set_default_aligner" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>set default aligner</div></blockquote>
<dl class="docutils">
<dt>set_default_aligner( (Result_factory)arg1, (str)name, (str)version [, (str)parameters=&#8217;&#8217; [, (str)uri=&#8217;&#8217; [, (int)run_id=0]]]) -&gt; Aligner_id :</dt>
<dd>set default aligner</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_factory.set_default_gl_database">
<tt class="descname">set_default_gl_database</tt><big>(</big><em>(Result_factory)arg1</em>, <em>(Gl_db_id)arg2</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Result_factory.set_default_gl_database" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>set default germline database</div></blockquote>
<dl class="docutils">
<dt>set_default_gl_database( (Result_factory)arg1, (str)name, (str)version, (str)species [, (str)url=&#8217;&#8216;]) -&gt; Gl_db_id :</dt>
<dd>set default germline database</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_factory.set_default_num_system">
<tt class="descname">set_default_num_system</tt><big>(</big><em>(Result_factory)arg1</em>, <em>(Numsys_id)arg2</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Result_factory.set_default_num_system" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>set default numbering system</div></blockquote>
<dl class="docutils">
<dt>set_default_num_system( (Result_factory)arg1, (str)name) -&gt; Numsys_id :</dt>
<dd>set default numbering system</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Result_store">
<em class="property">class </em><tt class="descname">Result_store</tt><big>(</big><em>(object)arg1</em><span class="optional">[</span>, <em>(Results_meta)meta=None</em><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Result_store" title="Permalink to this definition">¶</a></dt>
<dd><p>Storage of sequencing read results</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Result_store.empty">
<tt class="descname">empty</tt><big>(</big><em>(Result_store)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Result_store.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_store.insert">
<tt class="descname">insert</tt><big>(</big><em>(Result_store)arg1</em>, <em>(Read_result)arg2</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Result_store.insert" title="Permalink to this definition">¶</a></dt>
<dd><p>add new result</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Result_store.meta">
<tt class="descname">meta</tt><big>(</big><em>(Result_store)arg1</em><big>)</big> &rarr; Results_meta<a class="headerlink" href="#vdjml._vdjml_py.Result_store.meta" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Results_meta">
<em class="property">class </em><tt class="descname">Results_meta</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Results_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>Metadata for a collection of alignment results of sequencing reads</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Results_meta.aligner_map">
<tt class="descname">aligner_map</tt><big>(</big><em>(Results_meta)arg1</em><big>)</big> &rarr; Aligner_map :<a class="headerlink" href="#vdjml._vdjml_py.Results_meta.aligner_map" title="Permalink to this definition">¶</a></dt>
<dd><p>return a map of aligner software descriptions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Results_meta.gene_region_map">
<tt class="descname">gene_region_map</tt><big>(</big><em>(Results_meta)arg1</em><big>)</big> &rarr; Gene_region_map :<a class="headerlink" href="#vdjml._vdjml_py.Results_meta.gene_region_map" title="Permalink to this definition">¶</a></dt>
<dd><p>return a map of gene region descriptions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Results_meta.gl_db_map">
<tt class="descname">gl_db_map</tt><big>(</big><em>(Results_meta)arg1</em><big>)</big> &rarr; Germline_db_map :<a class="headerlink" href="#vdjml._vdjml_py.Results_meta.gl_db_map" title="Permalink to this definition">¶</a></dt>
<dd><p>return a map of germline database descriptions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Results_meta.gl_segment_map">
<tt class="descname">gl_segment_map</tt><big>(</big><em>(Results_meta)arg1</em><big>)</big> &rarr; Gl_segment_map :<a class="headerlink" href="#vdjml._vdjml_py.Results_meta.gl_segment_map" title="Permalink to this definition">¶</a></dt>
<dd><p>return a map of germline segment descriptions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Results_meta.insert">
<tt class="descname">insert</tt><big>(</big><em>(Results_meta)arg1</em>, <em>(Aligner_info)arg2</em><big>)</big> &rarr; Aligner_id :<a class="headerlink" href="#vdjml._vdjml_py.Results_meta.insert" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>insert information about aligner software</div></blockquote>
<dl class="docutils">
<dt>insert( (Results_meta)arg1, (Gl_db_info)arg2) -&gt; Gl_db_id :</dt>
<dd>insert information about database of germline segments</dd>
<dt>insert( (Results_meta)arg1, (Gl_segment_info)arg2) -&gt; Gl_seg_id :</dt>
<dd>insert information about germline segment</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Results_meta.num_system_map">
<tt class="descname">num_system_map</tt><big>(</big><em>(Results_meta)arg1</em><big>)</big> &rarr; Num_system_map :<a class="headerlink" href="#vdjml._vdjml_py.Results_meta.num_system_map" title="Permalink to this definition">¶</a></dt>
<dd><p>return a map of numbering systems</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Seg_match_id">
<em class="property">class </em><tt class="descname">Seg_match_id</tt><big>(</big><em>(object)arg1</em><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Seg_match_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Segment match ID</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Segment_combination">
<em class="property">class </em><tt class="descname">Segment_combination</tt><big>(</big><em>(object)arg1</em>, <em>(Seg_match_id)seg_match_1</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_2=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb9750&gt;</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_3=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb97c0&gt;</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_4=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb9830&gt;</em><span class="optional">[</span>, <em>(Seg_match_id)seg_match_5=&lt;vdjml._vdjml_py.Seg_match_id object at 0x7fa23ecb98a0&gt;</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Segment_combination" title="Permalink to this definition">¶</a></dt>
<dd><p>combination of aligned germline segments</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Segment_combination.insert">
<tt class="descname">insert</tt><big>(</big><em>(Segment_combination)arg1</em>, <em>(Seg_match_id)arg2</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Segment_combination.insert" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>insert segment match ID</div></blockquote>
<dl class="docutils">
<dt>insert( (Segment_combination)arg1, (Gene_region)arg2) -&gt; None :</dt>
<dd>insert gene region</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_combination.regions">
<tt class="descname">regions</tt><big>(</big><em>(Segment_combination)arg1</em><big>)</big> &rarr; Gene_region_set :<a class="headerlink" href="#vdjml._vdjml_py.Segment_combination.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>collection of gene regions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_combination.segments">
<tt class="descname">segments</tt><big>(</big><em>(Segment_combination)arg1</em><big>)</big> &rarr; Segment_match_id_set :<a class="headerlink" href="#vdjml._vdjml_py.Segment_combination.segments" title="Permalink to this definition">¶</a></dt>
<dd><p>set of segment match IDs</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Segment_combination_builder">
<em class="property">class </em><tt class="descname">Segment_combination_builder</tt><a class="headerlink" href="#vdjml._vdjml_py.Segment_combination_builder" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct alignment results for a combination of germline gene segments</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Segment_combination_builder.insert_region">
<tt class="descname">insert_region</tt><big>(</big><em>(Segment_combination_builder)arg1</em>, <em>(str)name</em>, <em>(Interval)read_range</em><span class="optional">[</span>, <em>(Match_metrics)metric=&lt;vdjml._vdjml_py.Match_metrics object at 0x7fa23ecc7398&gt;</em><span class="optional">[</span>, <em>(Numsys_id)num_system=&lt;vdjml._vdjml_py.Numsys_id object at 0x7fa23ecb9980&gt;</em><span class="optional">[</span>, <em>(Aligner_id)aligner_id=&lt;vdjml._vdjml_py.Aligner_id object at 0x7fa23ecb99f0&gt;</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; None<a class="headerlink" href="#vdjml._vdjml_py.Segment_combination_builder.insert_region" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>insert_region( (Segment_combination_builder)arg1, (Region_id)region, (Interval)read_range [, (Match_metrics)metric=&lt;vdjml._vdjml_py.Match_metrics object at 0x7fa23ecc7410&gt; [, (Numsys_id)num_system=&lt;vdjml._vdjml_py.Numsys_id object at 0x7fa23ecb9a60&gt; [, (Aligner_id)aligner_id=&lt;vdjml._vdjml_py.Aligner_id object at 0x7fa23ecb9ad0&gt;]]]) -&gt; None :</dt>
<dd><p class="first">indicate gene region location in read sequence</p>
<table class="last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">param region:</th><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Region_id" title="vdjml._vdjml_py.Region_id"><tt class="xref py py-class docutils literal"><span class="pre">Region_id</span></tt></a>, type of region</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">param read_range:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Interval" title="vdjml._vdjml_py.Interval"><tt class="xref py py-class docutils literal"><span class="pre">Interval</span></tt></a>, start and end positions in read sequence</td>
</tr>
<tr class="field-odd field"><th class="field-name">param metric:</th><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Match_metrics" title="vdjml._vdjml_py.Match_metrics"><tt class="xref py py-class docutils literal"><span class="pre">Match_metrics</span></tt></a>, alignment metrics between read and germline sequences; default: no metrics recorded</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">param num_system:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Numsys_id" title="vdjml._vdjml_py.Numsys_id"><tt class="xref py py-class docutils literal"><span class="pre">Numsys_id</span></tt></a>, default: no numbering system recorded</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">param aligner_id:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Aligner_id" title="vdjml._vdjml_py.Aligner_id"><tt class="xref py py-class docutils literal"><span class="pre">Aligner_id</span></tt></a>, default: current aligner ID is used</td>
</tr>
</tbody>
</table>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Segment_match">
<em class="property">class </em><tt class="descname">Segment_match</tt><big>(</big><em>(object)arg1</em>, <em>(int)read_pos0</em>, <em>(Btop)btop</em><span class="optional">[</span>, <em>(Match_metrics)match_metrics=&lt;vdjml._vdjml_py.Match_metrics object at 0x7fa23ef4c848&gt;</em><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Segment_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment results for a read segment</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.aa_substitutions">
<tt class="descname">aa_substitutions</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Aa_substitutions_set :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.aa_substitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>amino acid substitutions</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.btop">
<tt class="descname">btop</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Btop :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.btop" title="Permalink to this definition">¶</a></dt>
<dd><p>BTOP alignment description</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.gl_length">
<tt class="descname">gl_length</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.gl_length" title="Permalink to this definition">¶</a></dt>
<dd><p>length of the aligned germline segment(s)</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.gl_range">
<tt class="descname">gl_range</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Interval :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.gl_range" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>nucleotide range of the first germline sequence that matches the read sequence</div></blockquote>
<dl class="docutils">
<dt>gl_range( (Segment_match)arg1, (Gl_segment_match)arg2) -&gt; Interval :</dt>
<dd>nucleotide range of the specified germline sequence that matches the read sequence</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.gl_segments">
<tt class="descname">gl_segments</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Gl_segment_map :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.gl_segments" title="Permalink to this definition">¶</a></dt>
<dd><p>germline segment map</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.id">
<tt class="descname">id</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Seg_match_id :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.id" title="Permalink to this definition">¶</a></dt>
<dd><p>segment match ID</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.insert">
<tt class="descname">insert</tt><big>(</big><em>(Segment_match)arg1</em>, <em>(Gl_segment_match)arg2</em><big>)</big> &rarr; Gl_seg_match_id :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.insert" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>insert germline segment match</div></blockquote>
<dl class="docutils">
<dt>insert( (Segment_match)arg1, (Aa_substitution)arg2) -&gt; None :</dt>
<dd>insert amino acid substitution</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.match_metrics">
<tt class="descname">match_metrics</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Match_metrics :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.match_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>alignment metrics</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match.read_range">
<tt class="descname">read_range</tt><big>(</big><em>(Segment_match)arg1</em><big>)</big> &rarr; Interval :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match.read_range" title="Permalink to this definition">¶</a></dt>
<dd><p>read sequence nucleotide range that matches to germline segment</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Segment_match_builder">
<em class="property">class </em><tt class="descname">Segment_match_builder</tt><a class="headerlink" href="#vdjml._vdjml_py.Segment_match_builder" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct alignment results for one sequencing read segment match</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match_builder.get">
<tt class="descname">get</tt><big>(</big><em>(Segment_match_builder)arg1</em><big>)</big> &rarr; Segment_match :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match_builder.get" title="Permalink to this definition">¶</a></dt>
<dd><p>get segment match structure</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match_builder.insert_aa_substitution">
<tt class="descname">insert_aa_substitution</tt><big>(</big><em>(Segment_match_builder)arg1</em>, <em>(int)read_pos0</em>, <em>(str)read_aa</em>, <em>(str)gl_aa</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match_builder.insert_aa_substitution" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>add amino acid substitution information</div></blockquote>
<dl class="docutils">
<dt>insert_aa_substitution( (Segment_match_builder)arg1, (int)read_pos0, (str)read_aa, (str)gl_aa) -&gt; None :</dt>
<dd>add amino acid substitution information</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match_builder.insert_gl_segment_match">
<tt class="descname">insert_gl_segment_match</tt><big>(</big><em>(Segment_match_builder)arg1</em>, <em>(Gl_seg_id)gl_segment_id</em>, <em>(int)pos0</em><span class="optional">[</span>, <em>(Numsys_id)num_system_id=&lt;vdjml._vdjml_py.Numsys_id object at 0x7fa23ecb9b40&gt;</em><span class="optional">[</span>, <em>(Aligner_id)aligner=&lt;vdjml._vdjml_py.Aligner_id object at 0x7fa23ecb9bb0&gt;</em><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; Gl_seg_match_id :<a class="headerlink" href="#vdjml._vdjml_py.Segment_match_builder.insert_gl_segment_match" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>add germline segment alignment info</div></blockquote>
<dl class="docutils">
<dt>insert_gl_segment_match( (Segment_match_builder)arg1, (str)vdj, (str)seg_name, (int)gl_pos0 [, (Gl_db_id)gl_database=&lt;vdjml._vdjml_py.Gl_db_id object at 0x7fa23ecb9c20&gt; [, (Numsys_id)num_system=&lt;vdjml._vdjml_py.Numsys_id object at 0x7fa23ecb9c90&gt; [, (Aligner_id)aligner=&lt;vdjml._vdjml_py.Aligner_id object at 0x7fa23ecb9d00&gt;]]]) -&gt; Gl_seg_match_id :</dt>
<dd>add germline segment alignment info</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Segment_match_id_set">
<em class="property">class </em><tt class="descname">Segment_match_id_set</tt><a class="headerlink" href="#vdjml._vdjml_py.Segment_match_id_set" title="Permalink to this definition">¶</a></dt>
<dd><p>set of segment match IDs</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match_id_set.empty">
<tt class="descname">empty</tt><big>(</big><em>(Segment_match_id_set)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Segment_match_id_set.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Segment_match_map">
<em class="property">class </em><tt class="descname">Segment_match_map</tt><a class="headerlink" href="#vdjml._vdjml_py.Segment_match_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Collection of segment matches</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Segment_match_map.empty">
<tt class="descname">empty</tt><big>(</big><em>(Segment_match_map)arg1</em><big>)</big> &rarr; bool<a class="headerlink" href="#vdjml._vdjml_py.Segment_match_map.empty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Sequence_match">
<em class="property">class </em><tt class="descname">Sequence_match</tt><a class="headerlink" href="#vdjml._vdjml_py.Sequence_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligned sequences, start and end indices</p>
<dl class="attribute">
<dt id="vdjml._vdjml_py.Sequence_match.end_">
<tt class="descname">end_</tt><a class="headerlink" href="#vdjml._vdjml_py.Sequence_match.end_" title="Permalink to this definition">¶</a></dt>
<dd><p>last plus one position for aligned read and germline sequences</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Sequence_match.seq_">
<tt class="descname">seq_</tt><a class="headerlink" href="#vdjml._vdjml_py.Sequence_match.seq_" title="Permalink to this definition">¶</a></dt>
<dd><p>aligned read and germline sequences</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Sequence_match.start_">
<tt class="descname">start_</tt><a class="headerlink" href="#vdjml._vdjml_py.Sequence_match.start_" title="Permalink to this definition">¶</a></dt>
<dd><p>0-based starting position for aligned read and germline sequences</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Vdjml_generator_info">
<em class="property">class </em><tt class="descname">Vdjml_generator_info</tt><a class="headerlink" href="#vdjml._vdjml_py.Vdjml_generator_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Info about aligner software</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_generator_info.datetime">
<tt class="descname">datetime</tt><big>(</big><em>(Vdjml_generator_info)arg1</em><big>)</big> &rarr; object :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_generator_info.datetime" title="Permalink to this definition">¶</a></dt>
<dd><p>file creation date and time, GMT</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_generator_info.datetime_str">
<tt class="descname">datetime_str</tt><big>(</big><em>(Vdjml_generator_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_generator_info.datetime_str" title="Permalink to this definition">¶</a></dt>
<dd><p>file creation date and time, GMT</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_generator_info.name">
<tt class="descname">name</tt><big>(</big><em>(Vdjml_generator_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_generator_info.name" title="Permalink to this definition">¶</a></dt>
<dd><p>VDJML file generator name</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_generator_info.version">
<tt class="descname">version</tt><big>(</big><em>(Vdjml_generator_info)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_generator_info.version" title="Permalink to this definition">¶</a></dt>
<dd><p>VDJML file generator version</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Vdjml_reader">
<em class="property">class </em><tt class="descname">Vdjml_reader</tt><big>(</big><em>(object)arg1</em>, <em>(str)file_name</em><span class="optional">[</span>, <em>(Compression)compression=vdjml._vdjml_py.Compression.Uncompressed</em><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader" title="Permalink to this definition">¶</a></dt>
<dd><p>Incrementally parse VDJML read-by-read</p>
<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.generator_info">
<tt class="descname">generator_info</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; Vdjml_generator_info :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.generator_info" title="Permalink to this definition">¶</a></dt>
<dd><p>information about VDJML generator</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.has_result">
<tt class="descname">has_result</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; bool :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.has_result" title="Permalink to this definition">¶</a></dt>
<dd><p>return True if result was found</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.meta">
<tt class="descname">meta</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; Results_meta :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.meta" title="Permalink to this definition">¶</a></dt>
<dd><p>results meta</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.next">
<tt class="descname">next</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; None :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.next" title="Permalink to this definition">¶</a></dt>
<dd><p>parse next read result</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.result">
<tt class="descname">result</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; Read_result :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.result" title="Permalink to this definition">¶</a></dt>
<dd><p>return parsed result</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.version">
<tt class="descname">version</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; int :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.version" title="Permalink to this definition">¶</a></dt>
<dd><p>VDJML version of the file</p>
</dd></dl>

<dl class="method">
<dt id="vdjml._vdjml_py.Vdjml_reader.version_str">
<tt class="descname">version_str</tt><big>(</big><em>(Vdjml_reader)arg1</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.Vdjml_reader.version_str" title="Permalink to this definition">¶</a></dt>
<dd><p>VDJML version of the file</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Vdjml_writer">
<em class="property">class </em><tt class="descname">Vdjml_writer</tt><big>(</big><em>(object)arg1</em>, <em>(str)file_name</em>, <em>(Results_meta)meta</em><span class="optional">[</span>, <em>(Compression)compression=vdjml._vdjml_py.Compression.Unknown_compression</em><span class="optional">[</span>, <em>(int)version=1000</em><span class="optional">[</span>, <em>(Xml_writer_options)options=&lt;vdjml._vdjml_py.Xml_writer_options object at 0x7fa23efab9e0&gt;</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Vdjml_writer" title="Permalink to this definition">¶</a></dt>
<dd><p>Incrementally serialize VDJ alignment results</p>
</dd></dl>

<dl class="class">
<dt id="vdjml._vdjml_py.Xml_writer_options">
<em class="property">class </em><tt class="descname">Xml_writer_options</tt><big>(</big><em>(object)arg1</em><span class="optional">[</span>, <em>(str)indent='   '</em><span class="optional">[</span>, <em>(str)encoding='UTF-8'</em><span class="optional">[</span>, <em>(str)quote='&quot;'</em><span class="optional">[</span>, <em>(str)xml_version='1.0'</em><span class="optional">[</span>, <em>(int)buff_size=1024</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big><a class="headerlink" href="#vdjml._vdjml_py.Xml_writer_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Options for XML output</p>
<dl class="attribute">
<dt id="vdjml._vdjml_py.Xml_writer_options.buff_size">
<tt class="descname">buff_size</tt><a class="headerlink" href="#vdjml._vdjml_py.Xml_writer_options.buff_size" title="Permalink to this definition">¶</a></dt>
<dd><p>output buffer size</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Xml_writer_options.indent">
<tt class="descname">indent</tt><a class="headerlink" href="#vdjml._vdjml_py.Xml_writer_options.indent" title="Permalink to this definition">¶</a></dt>
<dd><p>indentation string</p>
</dd></dl>

<dl class="attribute">
<dt id="vdjml._vdjml_py.Xml_writer_options.quote">
<tt class="descname">quote</tt><a class="headerlink" href="#vdjml._vdjml_py.Xml_writer_options.quote" title="Permalink to this definition">¶</a></dt>
<dd><p>quotation character</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.codons">
<tt class="descname">codons</tt><big>(</big><em>(Btop)btop</em><span class="optional">[</span>, <em>(int)read_start=18446744073709551615L</em><span class="optional">[</span>, <em>(int)gl_start=18446744073709551615L</em><span class="optional">[</span>, <em>(str)read_seq=''</em><span class="optional">[</span>, <em>(str)gl_seq=''</em><span class="optional">[</span>, <em>(bool)follow_read=False</em><span class="optional">[</span>, <em>(bool)follow_gl=False</em><span class="optional">[</span>, <em>(str)match_char='.'</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; object :<a class="headerlink" href="#vdjml._vdjml_py.codons" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">codon iterator</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.mismatches">
<tt class="descname">mismatches</tt><big>(</big><em>(Btop)btop</em><big>)</big> &rarr; Mismatch_iter :<a class="headerlink" href="#vdjml._vdjml_py.mismatches" title="Permalink to this definition">¶</a></dt>
<dd><p>return nucleotide mismatch iterator</p>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.nucleotide_match">
<tt class="descname">nucleotide_match</tt><big>(</big><em>(Btop)btop</em><span class="optional">[</span>, <em>(int)read_pos0=18446744073709551615L</em><span class="optional">[</span>, <em>(int)gl_pos0=18446744073709551615L</em><span class="optional">[</span>, <em>(str)read_seq=''</em><span class="optional">[</span>, <em>(str)gl_seq=''</em><span class="optional">[</span>, <em>(str)match_char='.'</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; Nucleotide_match :<a class="headerlink" href="#vdjml._vdjml_py.nucleotide_match" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>Provides information about a pair of aligned nucleotides</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">param btop:</th><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Btop" title="vdjml._vdjml_py.Btop"><tt class="xref py py-class docutils literal"><span class="pre">Btop</span></tt></a> BTOP structure</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">param read_pos0:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">0-based position relative to read sequence</td>
</tr>
<tr class="field-odd field"><th class="field-name">param gl_pos0:</th><td class="field-body">0-based position relative to germline sequence</td>
</tr>
<tr class="field-even field"><th class="field-name">param read_seq:</th><td class="field-body">read sequence</td>
</tr>
<tr class="field-odd field"><th class="field-name">param gl_seq:</th><td class="field-body">germline sequence</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">param match_char:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><tt class="xref py py-class docutils literal"><span class="pre">char</span></tt>, character to indicate matching nucleotides</td>
</tr>
<tr class="field-odd field"><th class="field-name">return:</th><td class="field-body"><a class="reference internal" href="#vdjml._vdjml_py.Nucleotide_match" title="vdjml._vdjml_py.Nucleotide_match"><tt class="xref py py-class docutils literal"><span class="pre">Nucleotide_match</span></tt></a> information about two aligned nucleotides</td>
</tr>
</tbody>
</table>
</div></blockquote>
<dl class="docutils">
<dt>nucleotide_match( (Segment_match)sm, (object)pos0 [, (Gl_segment_match)gsm]) -&gt; Nucleotide_match :</dt>
<dd>Generate <a class="reference internal" href="#vdjml._vdjml_py.Nucleotide_match" title="vdjml._vdjml_py.Nucleotide_match"><tt class="xref py py-class docutils literal"><span class="pre">Nucleotide_match</span></tt></a>, information about two aligned</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.nucleotides">
<tt class="descname">nucleotides</tt><big>(</big><em>(Btop)btop</em><span class="optional">[</span>, <em>(str)read_seq=''</em><span class="optional">[</span>, <em>(str)gl_seq=''</em><span class="optional">[</span>, <em>(str)match_char='.'</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; object :<a class="headerlink" href="#vdjml._vdjml_py.nucleotides" title="Permalink to this definition">¶</a></dt>
<dd><p>return nucleotide iterator</p>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.numbering_system">
<tt class="descname">numbering_system</tt><big>(</big><em>(Gl_segment_match)gl_segment_match</em>, <em>(Results_meta)meta</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.numbering_system" title="Permalink to this definition">¶</a></dt>
<dd><p>return numbering system name</p>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.segment_name">
<tt class="descname">segment_name</tt><big>(</big><em>(Gl_segment_match)gl_segment_match</em>, <em>(Results_meta)meta</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.segment_name" title="Permalink to this definition">¶</a></dt>
<dd><p>return segment name</p>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.segment_type">
<tt class="descname">segment_type</tt><big>(</big><em>(Gl_segment_match)gl_segment_match</em>, <em>(Results_meta)meta</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.segment_type" title="Permalink to this definition">¶</a></dt>
<dd><p>return numbering system name</p>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.sequence_match">
<tt class="descname">sequence_match</tt><big>(</big><em>(Btop)btop</em><span class="optional">[</span>, <em>(int)read_start=18446744073709551615L</em><span class="optional">[</span>, <em>(int)read_end=18446744073709551615L</em><span class="optional">[</span>, <em>(int)gl_start=18446744073709551615L</em><span class="optional">[</span>, <em>(int)gl_end=18446744073709551615L</em><span class="optional">[</span>, <em>(str)read_seq=''</em><span class="optional">[</span>, <em>(str)gl_seq=''</em><span class="optional">[</span>, <em>(str)match_char='.'</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; Sequence_match :<a class="headerlink" href="#vdjml._vdjml_py.sequence_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a pair of aligned sequences with positions for start and end</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>btop</strong> &#8211; <a class="reference internal" href="#vdjml._vdjml_py.Btop" title="vdjml._vdjml_py.Btop"><tt class="xref py py-class docutils literal"><span class="pre">Btop</span></tt></a>, BTOP structure</li>
<li><strong>read_start</strong> &#8211; position for alignment start (0-based, relative to read sequence)</li>
<li><strong>read_end</strong> &#8211; position for alignment end (0-based, relative to read sequence)</li>
<li><strong>gl_start</strong> &#8211; position for alignment start (0-based, relative to germline sequence)</li>
<li><strong>gl_end</strong> &#8211; position for alignment end (0-based, relative to germline sequence)</li>
<li><strong>read_seq</strong> &#8211; read sequence</li>
<li><strong>gl_seq</strong> &#8211; germline sequence</li>
<li><strong>match_char</strong> &#8211; <tt class="xref py py-class docutils literal"><span class="pre">char</span></tt>, character to indicate matching nucleotides</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#vdjml._vdjml_py.Sequence_match" title="vdjml._vdjml_py.Sequence_match"><tt class="xref py py-class docutils literal"><span class="pre">Sequence_match</span></tt></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.trim_complement">
<tt class="descname">trim_complement</tt><big>(</big><em>(str)seq</em>, <em>(Interval)interval</em>, <em>(bool)reverse</em><big>)</big> &rarr; str :<a class="headerlink" href="#vdjml._vdjml_py.trim_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Trim and optionally reverse-complement a sequence</p>
</dd></dl>

<dl class="function">
<dt id="vdjml._vdjml_py.write_to_file">
<tt class="descname">write_to_file</tt><big>(</big><em>(str)path</em>, <em>(Result_store)store</em><span class="optional">[</span>, <em>(Compression)compression=vdjml._vdjml_py.Compression.Unknown_compression</em><span class="optional">[</span>, <em>(int)version=1000</em><span class="optional">[</span>, <em>(Xml_writer_options)options=&lt;vdjml._vdjml_py.Xml_writer_options object at 0x7fa23efab8c0&gt;</em><span class="optional">]</span><span class="optional">]</span><span class="optional">]</span><big>)</big> &rarr; None<a class="headerlink" href="#vdjml._vdjml_py.write_to_file" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-vdjml.igblast_parse"></span><p>vdjml/python/igblast_parse.py is part of VDJML project
Distributed under the Boost Software License, Version 1.0; see doc/license.txt.
Author Edward A. Salinas 2014</p>
<dl class="function">
<dt id="vdjml.igblast_parse.comp_dna">
<tt class="descname">comp_dna</tt><big>(</big><em>dna</em>, <em>allowIUPAC=False</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.comp_dna" title="Permalink to this definition">¶</a></dt>
<dd><p>complement a string (of DNA)
allow IUPAC complementing if desired</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.compareIGBlastJuncDataWithQueryJuncData">
<tt class="descname">compareIGBlastJuncDataWithQueryJuncData</tt><big>(</big><em>query_rec</em>, <em>igblastSeq</em>, <em>igBlastInterval</em>, <em>inverted_flag</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.compareIGBlastJuncDataWithQueryJuncData" title="Permalink to this definition">¶</a></dt>
<dd><p>The basic idea of this suburoutine is as follows:
1) Assuming the query read is given proceed to step #2
2) Recive the junction interval passed in (computed by the code) AND
receive the junction sequence passed in (given by IGBLAST)
3) compare the sequence given by IgBLAST with the sequence extracted from
the read (using the computed interval taking into account inversion or not)
4) if the SEQUENCE extracted from the read from the computed interval does
NOT match the sequence as given by IgBLAST, then print out an error message</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.extractAsItemOrFirstFromList">
<tt class="descname">extractAsItemOrFirstFromList</tt><big>(</big><em>t</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.extractAsItemOrFirstFromList" title="Permalink to this definition">¶</a></dt>
<dd><p>If an item is a list, return the first item if it&#8217;s found
if the item is not a list, just return the item</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.extractJunctionRegionSeq">
<tt class="descname">extractJunctionRegionSeq</tt><big>(</big><em>jRegion</em>, <em>query_rec</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.extractJunctionRegionSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>given a pyVDJML junction region, return the sequece as it would appear in
IGBLAST output</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.extractSubSeq">
<tt class="descname">extractSubSeq</tt><big>(</big><em>interval</em>, <em>seq</em>, <em>is_inverted</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.extractSubSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>given a seq record, an interval into it, and an inverted flag (telling
whether interval is in the opposite strand or not)
return the subsequence (inclusive) indicated by the interval
NOTE that the interval has 1-based indices</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.getRevCompInterval">
<tt class="descname">getRevCompInterval</tt><big>(</big><em>i</em>, <em>seq_len_in_bp</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.getRevCompInterval" title="Permalink to this definition">¶</a></dt>
<dd><p>given an interval (in list form [from,to])
with 1-based indesing AND given a sequence length in BP
return the same interval but on the reverse strand</p>
<p>form an interval in one strand
find the interval in the reverse complement strand</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.getSubstitutionsInsertionsDeletionsFromBTOP">
<tt class="descname">getSubstitutionsInsertionsDeletionsFromBTOP</tt><big>(</big><em>btop</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.getSubstitutionsInsertionsDeletionsFromBTOP" title="Permalink to this definition">¶</a></dt>
<dd><p>from a BTOP string extract the numbers of insertions, deletions, and
substitutions
in the BTOP for the pairs, the first character belongs to the READ(query),
the second to the GERMLINE(subject)</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.makeMap">
<tt class="descname">makeMap</tt><big>(</big><em>col_list</em>, <em>val_tab_str</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.makeMap" title="Permalink to this definition">¶</a></dt>
<dd><p>from a column list (keys) and tab-separated values
make a dict/map</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.makeMetricFromCharMap">
<tt class="descname">makeMetricFromCharMap</tt><big>(</big><em>charMap</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.makeMetricFromCharMap" title="Permalink to this definition">¶</a></dt>
<dd><p>given a characterization map for a region, make a metrics object</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.obtainJuncIntervalAndSeq">
<tt class="descname">obtainJuncIntervalAndSeq</tt><big>(</big><em>juncMapKey</em>, <em>juncMap</em>, <em>juncFirstStartSegMap</em>, <em>juncFirstEndSegMap</em>, <em>isVJJunc=False</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.obtainJuncIntervalAndSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>Analyze the junction sequence and look at the areas surrounding the junction 
(anchors on each end VJ, DJ, VD)
Based on the analysis compute a read interval and declare it to be the junction.
Return that interval as well as the sequence as a package via an array.
Testing of this code with real data (millions of reads) has shown that when 
intervals and sequences
are returned that when the interval is used with the actual read that the 
returned sequences from here
match the sequences retrieved from the read using the computed interval.</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.printMap">
<tt class="descname">printMap</tt><big>(</big><em>m</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.printMap" title="Permalink to this definition">¶</a></dt>
<dd><p>little utility to print a map</p>
</dd></dl>

<dl class="function">
<dt id="vdjml.igblast_parse.scanOutputToVDJML">
<tt class="descname">scanOutputToVDJML</tt><big>(</big><em>input_file</em>, <em>fact</em>, <em>fasta_query_path=None</em><big>)</big><a class="headerlink" href="#vdjml.igblast_parse.scanOutputToVDJML" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan lines of IgBLAST output
Use <cite>#</cite> at the beginning of lines to identify IgBLAST sections of output.
Based on those sections interpret/classify the output (as alignment summary or hit
data for example) and package/accumulate the data.
Then, once the end of a record is reached (as indicated by processing the
number of hits as it said it got)
Send the accumulated/aggreagated/packaged data to vdjml_read_serialize
to turn the data into a PyVDJML object. 
And then return that created/serialized object</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">VDJMLpy</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#module-vdjml">API reference</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter"><em>vdjml</em></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="build.html"
                        title="next chapter">Building <em>vdjml</em></a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="build.html" title="Building vdjml"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="vdjml"
             >previous</a> |</li>
        <li><a href="index.html">vdjml</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013-4, Mikhail K. Levin, Edward A. Salinas.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.1.
    </div>
  </body>
</html>