<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:vdj="http://vdjserver.org/vdjml/xsd/1/">
  <head>
    <title>VDJML | XML Schema</title>
    <meta http-equiv="content-type" content="text/html;charset=utf-8" />
    <link rel="stylesheet" href="xsd.css" type="text/css" />
  </head>
  <body><h1>VDJML XML Schema</h1>
    <p>VDJML is a file format for storing the results from VDJ analysis of 
    immune receptor (IR) sequence reads.
    The VDJ analysis includes aligning each input sequence read to 
    V, D, and J germline gene segments,
    and, based on the aligned segments, identifying various regions of 
    interest and other properties of the input sequence.
    </p>
    
    <p>The VDJ analysis is performed using either "raw" input sequence reads, 
    (e.g., as obtained from a sequencing instrument), 
    "consensus" read sequences, or sequences obtained from other sources.</p>
    
    <p>Representing the source and the nature of the input sequences is beyond 
    the scope of this file format.</p>
    
    <p>In this document, the input sequences used for the analysis will be 
    referred to as <i>read sequences</i> or <i>reads</i>.</p>
    
    <p>All elements and attributes defined by this version of the VDJML schema 
    have a namespace <code>http://vdjserver.org/vdjml/xsd/1/</code> 
    (prefix <code>vdj:</code>).
    The top element of the schema is <code>vdj:vdjml</code>.</p>
    
    <p>The schema uses primitive datatypes defined by the 
    <a href="http://www.w3.org/TR/xmlschema-2/">W3C XML Schema</a>.
    The primitive datatypes are defined in the namespace
    <code>http://www.w3.org/2001/XMLSchema</code>, prefix <code>xs:</code>, 
    e.g., <code>xs:decimal</code>.</p>
    
    <p>A VDJML file consists of two parts.
    General information about the analysis appears under the element 
    <a href="#element-meta"><code>vdj:meta</code></a> and the result for
    each sequence read appears in a sequence of 
    <a href="#element-read"><code>vdj:read</code></a> elements under the 
    <a href="#element-read_results"><code>vdj:read_results</code></a> element.</p>
    
    <p>The schema also allows some user-defined elements and attributes,
    which may appear under 
    <a href="#element-meta"><code>vdj:meta</code></a> and 
    <a href="#element-read"><code>vdj:read</code></a> elements.
    User-defined elements and attributes, should have namespaces other than
    <code>vdj</code>.</p>
   <p class="graphviz"><embed type="image/svg+xml" src="xsd_dot.svg" alt="VDJML Diagram" width="800"></embed></p><div id="xs_elements">
<div id="element-vdjml" class="element"><span class="element-name">vdjml</span>
      <div class="element-comment">
      Top element of a VDJML document. 
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="vdjml-attribute-version" class="attribute"><span class="attribute-name">version</span><span class="attribute-type">xs:decimal</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               version of the VDJML format, e.g., <code>1.0</code>
            </div>
         </div></div></div>
<div id="element-meta" class="element"><span class="element-name">meta</span>
      <div class="element-comment">
      Contains general information common to all analysis results.
      May also contain user-defined attributes and elements,
      which should use namespaces other than <code>vdj</code>.
      </div>
   </div>
<div id="element-generator" class="element"><span class="element-name">generator</span>
      <div class="element-comment">
      Describes the software that wrote the VDJML file
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="generator-attribute-name" class="attribute"><span class="attribute-name">name</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               The name of the VDJML generator, e.g., <code>libVDJML</code>
            </div>
         </div>
<div id="generator-attribute-version" class="attribute"><span class="attribute-name">version</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Version of the generator software, e.g., <code>1.42.0</code>
            </div>
         </div>
<div id="generator-attribute-time_gmt" class="attribute"><span class="attribute-name">time_gmt</span><span class="attribute-type">xs:dateTime</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Date and time of writing the file, GMT, in 
               <code>xs:dateTime</code> format, e.g., 
               <code>2014-07-24T14:47:24</code>
            </div>
         </div></div></div>
<div id="element-aligner" class="element"><span class="element-name">aligner</span>
      <div class="element-comment">
         Information about the aligner, a program that generated all or some
         of the results in the VDJML document by either performing the
         VDJ analysis de novo (e.g., IgBlast or VQuest) or by building upon
         the results of another aligner.
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="aligner-attribute-aligner_id" class="attribute"><span class="attribute-name">aligner_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               A unique identifier for the IR analysis software
            </div>
         </div>
<div id="aligner-attribute-name" class="attribute"><span class="attribute-name">name</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               The name of the aligner
            </div>
         </div>
<div id="aligner-attribute-version" class="attribute"><span class="attribute-name">version</span><span class="attribute-type">xs:string</span><span class="attribute-is-required"></span>
            <div class="attribute-comment">
               The version of the aligner
            </div>
         </div>
<div id="aligner-attribute-run_id" class="attribute"><span class="attribute-name">run_id</span><span class="attribute-type">xs:integer</span><span class="attribute-is-required"></span>
            <div class="attribute-comment">
               A unique identifier for the analysis run
            </div>
         </div>
<div id="aligner-attribute-uri" class="attribute"><span class="attribute-name">uri</span><span class="attribute-type">xs:anyURI</span><span class="attribute-is-required"></span>
            <div class="attribute-comment">
               A URL from which the aligner or information about it can be 
               obtained
            </div>
         </div></div>
<div id="element-parameters" class="element"><span class="element-name">parameters</span>
               <div class="element-comment">
                  Parameters used for running the VDJ analysis.
               </div>
            </div></div>
<div id="element-germline_db" class="element"><span class="element-name">germline_db</span>
      <div class="element-comment">
      Information about a germline database used for the analysis.
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="germline_db-attribute-gl_db_id" class="attribute"><span class="attribute-name">gl_db_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               A unique identifier for the germline database 
            </div>
         </div>
<div id="germline_db-attribute-name" class="attribute"><span class="attribute-name">name</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               The database name
            </div>
         </div>
<div id="germline_db-attribute-species" class="attribute"><span class="attribute-name">species</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               The germline database species
            </div>
         </div>
<div id="germline_db-attribute-version" class="attribute"><span class="attribute-name">version</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Germline database version
            </div>
         </div>
<div id="germline_db-attribute-uri" class="attribute"><span class="attribute-name">uri</span><span class="attribute-type">xs:anyURI</span><span class="attribute-is-required"></span>
            <div class="attribute-comment">
               The URL where the database or information about it can be 
               obtained
            </div>
         </div></div></div>
<div id="element-read_results" class="element"><span class="element-name">read_results</span>
      <div class="element-comment">
      Sequence of alignment results for individual sequence reads.
      It is recommended that the sequence read results appear in the same order
      as in the FASTA/Q files from which the alignment results were obtained.
      </div>
   </div>
<div id="element-read" class="element"><span class="element-name">read</span>
      <div class="element-comment">
      Stores the analysis results of a single sequence read.
      
      May also contain user-defined elements and attributes.
      The user-defined elements and attributes should be in namespaces other 
      than <code>vdj</code>.
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="read-attribute-read_id" class="attribute"><span class="attribute-name">read_id</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Unique identifier of the read sequence; 
               should be same as the sequence ID from the original FASTA/Q
               file
            </div>
         </div></div></div>
<div id="element-vdj_alignment" class="element"><span class="element-name">vdj_alignment</span>
      <div class="element-comment">
      Immune receptor sequence alignment results for the sequence read.
      </div>
   </div>
<div id="element-segment_match" class="element"><span class="element-name">segment_match</span>
      <div class="element-comment">
      Information about germline gene segment(s) aligned to the read sequence
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="segment_match-attribute-segment_match_id" class="attribute"><span class="attribute-name">segment_match_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
	         <div class="attribute-comment">
	         A unique identifier for segment match to be used in other parts 
               of the VDJML document
	         </div>
	      </div>
<div id="segment_match-attribute-read_pos0" class="attribute"><span class="attribute-name">read_pos0</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
	         <div class="attribute-comment">
	         Zero-based starting position of the match in the read sequence
	         </div>
	      </div>
<div id="segment_match-attribute-read_len" class="attribute"><span class="attribute-name">read_len</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
	         <div class="attribute-comment">
	         Number of nucleotides in the read sequence for the matching region
	         </div>
	      </div>
<div id="segment_match-attribute-gl_len" class="attribute"><span class="attribute-name">gl_len</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
	         <div class="attribute-comment">
	         length of aligned germline segment
	         </div>
	      </div>
<div id="segment_match-attribute-identity" class="attribute"><span class="attribute-name">identity</span><span class="attribute-type">vdj:Percent</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         Percent of nucleotide sequence identity, e.g., <code>90%</code>.
         </div>
      </div>
<div id="segment_match-attribute-score" class="attribute"><span class="attribute-name">score</span><span class="attribute-type">xs:integer</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         Alignment score, as defined by the aligner software
         </div>
      </div>
<div id="segment_match-attribute-insertions" class="attribute"><span class="attribute-name">insertions</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         number of nucleotide insertions in the read sequence aligned to
         the germline sequence
         </div>
      </div>
<div id="segment_match-attribute-deletions" class="attribute"><span class="attribute-name">deletions</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         number of nucleotide deletions from the read sequence aligned to
         the germline sequence
         </div>
      </div>
<div id="segment_match-attribute-substitutions" class="attribute"><span class="attribute-name">substitutions</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         number of nucleotide substitutions
         </div>
      </div>
<div id="segment_match-attribute-stop_codon" class="attribute"><span class="attribute-name">stop_codon</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if the stop codon is present
         </div>
      </div>
<div id="segment_match-attribute-mutated_invariant" class="attribute"><span class="attribute-name">mutated_invariant</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if a codon for a conserved amino acid is mutated
         </div>
      </div>
<div id="segment_match-attribute-inverted" class="attribute"><span class="attribute-name">inverted</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if the read sequence is a reverse-complement
         to germline gene segments
         </div>
      </div>
<div id="segment_match-attribute-out_frame_indel" class="attribute"><span class="attribute-name">out_frame_indel</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if indel mutation resulted in a frame shift
         </div>
      </div>
<div id="segment_match-attribute-out_frame_vdj" class="attribute"><span class="attribute-name">out_frame_vdj</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if V(D)J recombination occurred out of frame
         </div>
      </div></div></div>
<div id="element-btop" class="element"><span class="element-name">btop</span>
      <div class="element-comment">
      BTOP (BLAST traceback operations) string.
      Integer indicates the number of aligned nucleotides. 
      A pair of letters indicates a nucleotide mismatch between read/germline 
      sequences.
      Dash ('-') indicates deletion.
      Example: <code>5AC-G35</code>.
      </div>
   </div>
<div id="element-gl_seg_match" class="element"><span class="element-name">gl_seg_match</span>
      <div class="element-comment">
         Information about aligned germline segment
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="gl_seg_match-attribute-gl_seg_match_id" class="attribute"><span class="attribute-name">gl_seg_match_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               A unique identifier for germline segment match to be used in 
               other parts of the VDJML document
            </div>
         </div>
<div id="gl_seg_match-attribute-num_system" class="attribute"><span class="attribute-name">num_system</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Numbering system name.
            </div>
         </div>
<div id="gl_seg_match-attribute-name" class="attribute"><span class="attribute-name">name</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Name of the germline segment
            </div>
         </div>
<div id="gl_seg_match-attribute-gl_db_id" class="attribute"><span class="attribute-name">gl_db_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               A unique identifier referring to the 
               <a href="#germline_db-attribute-gl_db_id">germline database</a>
               in which the segment was found
            </div>
         </div>
<div id="gl_seg_match-attribute-aligner_id" class="attribute"><span class="attribute-name">aligner_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               A unique identifier referring to the 
               <a href="aligner-attribute-aligner_id">aligner software</a>
               that produced the alignment
            </div>
         </div>
<div id="gl_seg_match-attribute-type" class="attribute"><span class="attribute-name">type</span><span class="attribute-type">vdj:Segment_type</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               The type of the segment 
               (<code>V</code>, <code>D</code>, or <code>J</code>)
            </div>
         </div>
<div id="gl_seg_match-attribute-gl_pos0" class="attribute"><span class="attribute-name">gl_pos0</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Zero-based start of the alignment position in the germline 
               sequence.
            </div>
         </div></div></div>
<div id="element-aa_substitution" class="element"><span class="element-name">aa_substitution</span>
      <div class="element-comment">
         Information about an amino acid codon substitution
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="aa_substitution-attribute-read_pos0" class="attribute"><span class="attribute-name">read_pos0</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Zero-based position of the first nucleotide of the codon in
               the read sequence.
            </div>
         </div>
<div id="aa_substitution-attribute-read_aa" class="attribute"><span class="attribute-name">read_aa</span><span class="attribute-type">vdj:Aminoacid</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               amino acid encoded by the read sequence
            </div>
         </div>
<div id="aa_substitution-attribute-gl_aa" class="attribute"><span class="attribute-name">gl_aa</span><span class="attribute-type">vdj:Aminoacid</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               amino acid encoded by the germline sequence
            </div>
         </div></div></div>
<div id="element-combination" class="element"><span class="element-name">combination</span>
      <div class="element-comment">
         Annotations of the read sequence based on certain alignments
         with germline gene segments.
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="combination-attribute-segments" class="attribute"><span class="attribute-name">segments</span><span class="attribute-type"></span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               List of identifiers for the segment matches 
               (<a href="#segment_match-attribute-segment_match_id"><code>segment_match_id</code></a>-s)
               that serve as a basis for the annotations listed in this element.
            </div>
         </div></div></div>
<div id="element-region" class="element"><span class="element-name">region</span>
      <div class="element-comment">
      Region of an immune receptor gene, i.e., Ig domain, segment junction.
      Note: the corresponding positions of the germline sequences are not stored
      because they may differ depending on the individual germline segment
      matches.
      </div>
   <div class="element-attributes"><div class="attribute-header"><span>attribute</span><span>type</span><span>required</span><span>description</span></div>
<div id="region-attribute-name" class="attribute"><span class="attribute-name">name</span><span class="attribute-type">xs:string</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Name of the region as provided by the alignment software,
               e.g., <code>FR1</code>
            </div>
         </div>
<div id="region-attribute-num_system" class="attribute"><span class="attribute-name">num_system</span><span class="attribute-type">xs:string</span><span class="attribute-is-required"></span>
            <div class="attribute-comment">
               Numbering system name.
               Attribute is not required because for some regions, e.g., 
               for VD-junctions, numbering system is not applicable.
            </div>
         </div>
<div id="region-attribute-aligner_id" class="attribute"><span class="attribute-name">aligner_id</span><span class="attribute-type">xs:positiveInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Unique identifier for the IR analysis software that produced the 
               region annotation 
               <a href="#aligner-attribute-aligner_id"><code>aligner_id</code></a>
            </div>
         </div>
<div id="region-attribute-read_pos0" class="attribute"><span class="attribute-name">read_pos0</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Zero-based index of the first read nucleotide that is mapped to
               the region.
               In case of an Ig domain, the corresponding position
               for each germline segment match may be determined based on 
               segment match positions and BTOPs.
            </div>
         </div>
<div id="region-attribute-read_len" class="attribute"><span class="attribute-name">read_len</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required">✓</span>
            <div class="attribute-comment">
               Length of the read sequence that corresponds to the region.
            </div>
         </div>
<div id="region-attribute-identity" class="attribute"><span class="attribute-name">identity</span><span class="attribute-type">vdj:Percent</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         Percent of nucleotide sequence identity, e.g., <code>90%</code>.
         </div>
      </div>
<div id="region-attribute-score" class="attribute"><span class="attribute-name">score</span><span class="attribute-type">xs:integer</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         Alignment score, as defined by the aligner software
         </div>
      </div>
<div id="region-attribute-insertions" class="attribute"><span class="attribute-name">insertions</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         number of nucleotide insertions in the read sequence aligned to
         the germline sequence
         </div>
      </div>
<div id="region-attribute-deletions" class="attribute"><span class="attribute-name">deletions</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         number of nucleotide deletions from the read sequence aligned to
         the germline sequence
         </div>
      </div>
<div id="region-attribute-substitutions" class="attribute"><span class="attribute-name">substitutions</span><span class="attribute-type">xs:nonNegativeInteger</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         number of nucleotide substitutions
         </div>
      </div>
<div id="region-attribute-stop_codon" class="attribute"><span class="attribute-name">stop_codon</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if the stop codon is present
         </div>
      </div>
<div id="region-attribute-mutated_invariant" class="attribute"><span class="attribute-name">mutated_invariant</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if a codon for a conserved amino acid is mutated
         </div>
      </div>
<div id="region-attribute-inverted" class="attribute"><span class="attribute-name">inverted</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if the read sequence is a reverse-complement
         to germline gene segments
         </div>
      </div>
<div id="region-attribute-out_frame_indel" class="attribute"><span class="attribute-name">out_frame_indel</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if indel mutation resulted in a frame shift
         </div>
      </div>
<div id="region-attribute-out_frame_vdj" class="attribute"><span class="attribute-name">out_frame_vdj</span><span class="attribute-type">xs:boolean</span><span class="attribute-is-required"></span>
         <div class="attribute-comment">
         <code>true</code> if V(D)J recombination occurred out of frame
         </div>
      </div></div></div></div></body>
</html>
